tutorials/07-resource-compose/contracts/BestPizzaPlace.cdc

```diff
@@ -0,0 +1,142 @@
+        }
+    }
+
+    pub fun createTopping(name: String, addBeforeBaking: Bool): @Topping {
+        return <-create Topping(name: name, addBeforeBaking: addBeforeBaking)
+    }
+
+    pub enum Grain: UInt8 {
+      pub case wheat
+      pub case rye
+      pub case spelt
+    }
+
+    pub resource Dough: Ingredient {
+        pub let name: String
+        pub let timeToBake: UInt
+
+        init(name: String, timeToBake: UInt) {
+            self.name = name
+            self.timeToBake = timeToBake
+        }
+    }
+
+    pub fun createDough(grain: Grain, timeToBake: UInt): @Dough {
+        let doughNames = {Grain.wheat: "wheat dough", Grain.rye : "rye dough", Grain.spelt : "spelt dough"}
+        let name: String = doughNames[grain]!
+        return <-create Dough(name: name, timeToBake: timeToBake)
+    }
+
+    pub resource Pizza {
+        pub let name: String
+        pub let dough: @Dough
+        pub let sauce: @Sauce
+        pub let toppings: @[Topping]
+
+        init(name: String, dough: @Dough, sauce: @Sauce) {
+            self.name = name
+            self.dough <- dough
+            self.sauce <- sauce
+            self.toppings <- []
+        }
+
+        pub fun addTopping(topping: @Topping) {
+           self.toppings.append(<- topping)
+        }
+
+        pub fun showOrder(): Order {
+            let toppingNames: [String] = []
+            //iteration is a bit special because of https://github.com/onflow/cadence/issues/704
+            var i = 0
+            while i < self.toppings.length {
+              let toppingRef = &self.toppings[i] as &Topping
+              toppingNames.append(toppingRef.name)
+              i = i + 1
+            }
+            return Order(pizzaName: self.name, dough: self.dough.name, timeToBake: self.dough.timeToBake, sauceType: self.sauce.name, spiciness: self.sauce.getSpiciness(), toppings: toppingNames)
+        }
+
+        destroy() {
+            destroy self.dough
+            destroy self.sauce
+            destroy self.toppings
+        }
+    }
+
+    pub fun createPizza(name: String, dough: @Dough, sauce: @Sauce): @Pizza {
+        return <-create Pizza(name: name, dough: <-dough, sauce: <-sauce)
+    }
+
+    // used to inform about the pizza ingredients
+    pub struct Order {
+        pub var pizzaName: String
+        pub var dough: String
+        pub var timeToBake: UInt
+        pub var sauceType: String
+        pub var spiciness: String
+        pub var toppings: [String]
+
+        init(pizzaName: String, dough: String, timeToBake: UInt, sauceType: String, spiciness: String, toppings: [String]) {
+            self.pizzaName = pizzaName
+            self.dough = dough
+            self.timeToBake = timeToBake
+            self.sauceType = sauceType
+            self.spiciness = spiciness
+            self.toppings = toppings
+        }
+    }
+
+    init() {
+        self.PizzaStoragePath = /storage/pizza
+        self.PizzaPublicPath = /public/pizza
```

Comment:
I would create some more specific names for these

Reply:
I changed those now.